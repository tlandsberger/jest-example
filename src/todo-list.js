const isSameDate = (l, r) => {
  return l.getFullYear() === r.getFullYear()
    && l.getMonth() === r.getMonth()
    && l.getDate() === r.getDate()
}

const isBeforeDate = (l, r) => {
  return l.getFullYear() < r.getFullYear()
    || l.getFullYear() === r.getFullYear() && l.getMonth() < r.getMonth()
    || l.getFullYear() === r.getFullYear() && l.getMonth() === r.getMonth() && l.getDate() < r.getDate();
}

export class TodoList {
  renderTarget;
  todos;

  constructor(renderTarget) {
    this.renderTarget = renderTarget;
    this.todos = [];
  }

  addTodo(title, dueDate) {
    if (!title)
      throw new Error("You must enter a title");
    if (isNaN(dueDate?.getTime()))
      throw new Error("You must enter a due date");

    this.todos.push({ title, dueDate });
    this.todos.sort((l, r) => l.dueDate - r.dueDate)
    this.renderTodos();
  }

  removeTodo(todo) {
    this.todos = this.todos.filter(e => e !== todo);
    this.renderTodos();
  }

  renderTodos() {
    this.renderTarget.replaceChildren(...this.todos.map(todo => {
      const title = document.createElement("span");
      title.appendChild(document.createTextNode(todo.title));

      const dueDate = document.createElement("span");
      dueDate.appendChild(document.createTextNode(
        `due on: ${todo.dueDate.toLocaleDateString("en-US")}`));

      const btnRemove = document.createElement("button");
      btnRemove.appendChild(
        document.createTextNode("Remove TODO"));
      btnRemove.onclick = () => this.removeTodo(todo);

      const container = document.createElement("div");
      container.appendChild(title);
      container.appendChild(dueDate);
      container.appendChild(btnRemove);
      const now = new Date();
      if (isBeforeDate(todo.dueDate, now)) {
        container.classList.add("overdue");
      } else if (isSameDate(todo.dueDate, now)) {
        container.classList.add("due");
      }

      return container;
    }));
  }
}
