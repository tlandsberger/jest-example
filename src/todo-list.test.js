import { TestWatcher } from '@jest/core';
import { jest } from '@jest/globals';
import { TodoList } from "./todo-list.js";

const expectOrderedDates = (todos) => {
  const timestamps = todos.map(e => e.dueDate.getTime());
  for (let i = 0; i < timestamps.length - 1; i++) {
    expect(timestamps[i]).toBeLessThanOrEqual(timestamps[i + 1]);
  }
}

describe("TodoList", () => {
  let todoList;

  beforeEach(() => {
    todoList = new TodoList(document.body);
    todoList.addTodo("send seminar paper to peer reviewers",
      new Date("2021-10-18"));
    todoList.addTodo("upload seminar paper to Moodle",
      new Date("2021-11-15"));
    todoList.addTodo("send presentation to peer reviewers",
      new Date("2021-12-06"));
    todoList.addTodo("upload presentation to Moodle",
      new Date("2021-12-22"));
  });

  it("should be created", () => {
    expect(todoList).toBeDefined();
  });

  describe("addTodo()", () => {
    beforeEach(() => {
      jest.spyOn(todoList, "renderTodos").mockReset();
    });

    it("should insert the added todo to the todo list", () => {
      const oldLength = todoList.todos.length;
      todoList.addTodo("do the shopping", new Date("2021-10-01"));
      expect(todoList.todos.length).toBeGreaterThan(oldLength);
      expect(todoList.todos).toContainEqual({
        title: "do the shopping",
        dueDate: new Date("2021-10-01")
      });
    });

    it("should sort the todos ascendening by due date", () => {
      todoList.addTodo("do the shopping", new Date("2021-10-01"));
      expectOrderedDates(todoList.todos);
    });

    it("should call renderTodos()", () => {
      todoList.addTodo("do the shopping", new Date("2021-10-01"));
      expect(todoList.renderTodos).toHaveBeenCalled();
    });

    it("should throw if no title is given", () => {
      expect(() => todoList.addTodo(undefined, new Date())).toThrow();
      expect(() => todoList.addTodo(null, new Date())).toThrow();
      expect(() => todoList.addTodo("", new Date())).toThrow();
    });

    it("should throw if an invalid date is given", () => {
      expect(() => todoList.addTodo("foo", undefined)).toThrow();
      expect(() => todoList.addTodo("foo", null)).toThrow();
      expect(() => todoList.addTodo("foo", new Date("invalid"))).toThrow();
    });
  });

  describe("removeTodo()", () => {
    beforeEach(() => {
      jest.spyOn(todoList, "renderTodos").mockReset();
    });

    it("should remove the given todo from the todo list", () => {
      const oldLength = todoList.todos.length;
      todoList.removeTodo(todoList.todos[0]);
      expect(todoList.todos.length).toBeLessThan(oldLength);
      expect(todoList.todos).not.toContainEqual({
        title: "send seminar paper to peer reviewers",
        dueDate: new Date("2021-10-18")
      });
    });

    it("should sort the todos ascendening by due date", () => {
      todoList.removeTodo(todoList.todos[0]);
      expectOrderedDates(todoList.todos);
    });

    it("should call renderTodos()", () => {
      todoList.removeTodo(todoList.todos[0]);
      expect(todoList.renderTodos).toHaveBeenCalled();
    });
  });

  describe("renderTodos()", () => {
    it("should render the todos (snapshot test)", () => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date("2021-11-15"));
      todoList.renderTodos();
      expect(document.body).toMatchSnapshot();
    });
  });
});
