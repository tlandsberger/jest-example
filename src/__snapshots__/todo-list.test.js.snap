// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renderTodos() should render the todos (snapshot test) 1`] = `
<body>
  <div
    class="overdue"
  >
    <span>
      send seminar paper to peer reviewers
    </span>
    <span>
      due on: 10/18/2021
    </span>
    <button>
      Remove TODO
    </button>
  </div>
  <div
    class="due"
  >
    <span>
      upload seminar paper to Moodle
    </span>
    <span>
      due on: 11/15/2021
    </span>
    <button>
      Remove TODO
    </button>
  </div>
  <div>
    <span>
      send presentation to peer reviewers
    </span>
    <span>
      due on: 12/6/2021
    </span>
    <button>
      Remove TODO
    </button>
  </div>
  <div>
    <span>
      upload presentation to Moodle
    </span>
    <span>
      due on: 12/22/2021
    </span>
    <button>
      Remove TODO
    </button>
  </div>
</body>
`;
